R/process_observers.R

```diff
@@ -231,106 +323,94 @@ propogate_observers_across_stops <- function(mbbs_county) {
-      print(paste(
-        "ERROR!",
-        na_rows$year,
-        "route",
-        na_rows$route_num,
-        "has only NA values for observers and no corrected record in mbbs_survey_events. Likely source of error: the ebird entry for stop 1 is missing observer information."
-      ))
+    # confirm that the passed_na_row passed is an NA, if it's not just return and exit this function
+    if (is.na(passed_na_row$observers) == FALSE) {
+      # return("observers not NA") #for testing
+      return(FALSE) # return is correct to use, exit function early.
+    }
+
+    # since the observer of this rocombo is NA,
+    # evaluate if it should be ignored (we're not missing data)
+    # or should throw an error (we're missing data)
+
+    # filter to the na rows in mbbs_county that this rocombo represents
+    na_rows <- mbbs_county %>% # take county df
+      filter(is.na(observers) == TRUE) %>% # filter to the NA rows
+      filter(route_num == passed_na_row$route_num) %>% # filter to the NA rows for this route
+      anti_join(survey_list, join_by(mbbs_county, route_num, year))
+    # only keep any rows with NA observers if that route/year combo is not already
+    # represented on the survey list.
+    # Years where observers == NA, but that are on the survey_list
+    # (and so have an observer) are cut.
+
+    # evaluate if this route has all it's observers on the survey_list
+    if (nrow(na_rows) == 0) {
+      # the NA observer seen on this passed_na_row has been corrected elsewhere,
+      # it's on the survey_list.
+      # return("observers corrected elsewhere") #for testing
+      return(FALSE)
+    } else {
+      # this NA observer is not already on the survey list. This is likely because
+      # (1). It's a new year of data, observers did not propagate to all the rows,
+      # and survey_list gets updated after observer_conversion_table.
+      # (2). It's a year of data where it's not on the survey_list and ALL the
+      # route's 'observers' column are NA
+
+      # filter mbbs_county to evaluate if ANY row of data from this route/year
+      # combo contains an observers value
+      # (and will therefore have been caught in a non-NA rowcombos[i])
+      non_na_rows <- mbbs_county %>% # take the county df
+        filter(year %in% na_rows$year) %>% # filter to the year we've got NA
+        filter(route_num == passed_na_row$route_num) %>% # filter to the route
+        filter(!is.na(observers)) # filter to any rows where obs is NOT NA
+
+      # if there are NO rows in the mbbs where this route/year combo
+      # has a non-NA observer, flag the error
+      if (nrow(non_na_rows) == 0) {
+        print(paste(
+          "ERROR!",
+          na_rows$year,
+          "route",
+          na_rows$route_num,
+          "has only NA values for observers and no corrected record in mbbs_survey_events. Likely source of error: the ebird entry for stop 1 is missing observer information."
+        ))
+        return(TRUE)
```

Comment:
The return() calls in this function are used correctly b/c they're used to exit from the function early